File: repro-agent/src/prompt.ts
export interface VerificationTarget {
  metric: string;
  expected_value: number;
  source: string;
  conditions: string;
}

export interface SignalEntry {
  value: boolean;
  evidence: string;
}

export interface BenchmarkEntry {
  custom_id: string;
  central_claim: string;
  claim_evidence: string;
  mre_config: string;
  web_verification: string;
  verified_links: {
    paper_or_project?: string[];
    code?: string[];
    dataset?: string[];
    weights?: string[];
  };
  signals: {
    code_available: SignalEntry;
    dataset_available: SignalEntry;
    weights_available: SignalEntry;
    dataset_is_standard: SignalEntry;
  };
  verification_targets: VerificationTarget[];
  agent_task: string;
  h100_hours_estimate: number;
  h100_estimate_basis: string;
  score?: number;
  tier?: string;
}

export const SYSTEM_PROMPT = `You are a reproduction agent for an ML-paper benchmark.

You are given a benchmark entry with a Minimal Reproduction Example (MRE) to run.
Follow the agent_task steps in order:

  Step 1 — Browse: use github_browse, hf_browse, or fetch_url to read the repository
             README, key source files, and any setup instructions.
  Step 2 — Setup: use bash to clone the repo and install dependencies.
  Step 3 — Run: use bash to execute the experiment as specified in mre_config.
  Step 4 — Report: print a single JSON object (no prose) with these keys:
             reproduction_status: "success" | "partial" | "failed"
             metric_results: [{ metric: <name>, actual_value: <number|null> }]
             claim_supported: true | false | null
             claim_assessment: <one paragraph — assess the overall pattern of results
               against the central_claim, not each number in isolation>
             failure_reason: <if failed, why>

The metric names in metric_results must match verification_targets[].metric exactly.
Do not skip steps. Do not fabricate results — only report values you observed from
actual command output.`;

export function buildPrompt(entry: BenchmarkEntry): string {
  const availability = [
    `code_available: ${entry.signals.code_available.value} — ${entry.signals.code_available.evidence}`,
    `dataset_available: ${entry.signals.dataset_available.value} — ${entry.signals.dataset_available.evidence}`,
    `weights_available: ${entry.signals.weights_available.value} — ${entry.signals.weights_available.evidence}`,
    `dataset_is_standard: ${entry.signals.dataset_is_standard.value} — ${entry.signals.dataset_is_standard.evidence}`,
  ].join("\n");

  const parts: string[] = [
    `# Paper: ${entry.custom_id}`,
    `## Central Claim\n${entry.central_claim}`,
    `## Claim Evidence\n${entry.claim_evidence}`,
    `## Artifact Availability\n${availability}`,
    `## Web Verification Notes\n${entry.web_verification}`,
    `## Verified Links\n${JSON.stringify(entry.verified_links, null, 2)}`,
    `## MRE Configuration\n${entry.mre_config}`,
    `## Estimated Compute\n${entry.h100_hours_estimate} H100-hours — ${entry.h100_estimate_basis}`,
    `## Verification Targets\n${JSON.stringify(entry.verification_targets, null, 2)}`,
    `## Agent Task\n${entry.agent_task}`,
  ];

  if (entry.score !== undefined || entry.tier !== undefined) {
    parts.push(`## Difficulty\nscore: ${entry.score ?? "n/a"}, tier: ${entry.tier ?? "n/a"}`);
  }

  return parts.join("\n\n");
}

File: src/reprocli_agent/prompt.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .phases import Phase
    from .state import ReproState

AGENT_PREAMBLE = """\
You are a reproduction agent for ML papers running on an HPC cluster.
You work in phases controlled by a harness. Each phase has a single, specific goal.

Rules:
- Follow the phase instructions exactly. Do not jump ahead to a later phase.
- Only call the tools listed as "Allowed tools" for the current phase.
- Write all required artifacts using write_file before stopping.
- Do NOT emit the final reproduction JSON unless you are in the finalize phase.
- If a repair is needed (e.g. missing package, build failure), fix it and retry
  within the current phase — the harness will route you back if needed.
- Work entirely inside the sandbox working directory."""


@dataclass
class SignalEntry:
    value: bool
    evidence: str


@dataclass
class VerificationTarget:
    metric: str
    expected_value: float
    source: str
    conditions: str


@dataclass
class BenchmarkEntry:
    custom_id: str
    central_claim: str
    claim_evidence: str
    mre_config: str
    web_verification: str
    verified_links: dict[str, list[str]]
    signals: dict[str, SignalEntry]
    verification_targets: list[VerificationTarget]
    agent_task: str
    h100_hours_estimate: float
    h100_estimate_basis: str
    score: int | None = None
    tier: str | None = None

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> BenchmarkEntry:
        signals = {
            k: SignalEntry(value=v["value"], evidence=v["evidence"])
            for k, v in d.get("signals", {}).items()
        }
        targets = [
            VerificationTarget(
                metric=t["metric"],
                expected_value=t["expected_value"],
                source=t["source"],
                conditions=t["conditions"],
            )
            for t in d.get("verification_targets", [])
        ]
        return cls(
            custom_id=d["custom_id"],
            central_claim=d["central_claim"],
            claim_evidence=d.get("claim_evidence", ""),
            mre_config=d["mre_config"],
            web_verification=d.get("web_verification", ""),
            verified_links=d.get("verified_links", {}),
            signals=signals,
            verification_targets=targets,
            agent_task=d["agent_task"],
            h100_hours_estimate=d.get("h100_hours_estimate", 0.0),
            h100_estimate_basis=d.get("h100_estimate_basis", ""),
            score=d.get("score"),
            tier=d.get("tier"),
        )


def build_prompt(entry: BenchmarkEntry) -> str:
    availability = "\n".join(
        f"{k}: {s.value} — {s.evidence}" for k, s in entry.signals.items()
    )
    parts = [
        f"# Paper: {entry.custom_id}",
        f"## Central Claim\n{entry.central_claim}",
        f"## Claim Evidence\n{entry.claim_evidence}",
        f"## Artifact Availability\n{availability}",
        f"## Web Verification Notes\n{entry.web_verification}",
        f"## Verified Links\n{json.dumps(entry.verified_links, indent=2)}",
        f"## MRE Configuration\n{entry.mre_config}",
        f"## Estimated Compute\n{entry.h100_hours_estimate} H100-hours — {entry.h100_estimate_basis}",
        f"## Verification Targets\n{json.dumps([t.__dict__ for t in entry.verification_targets], indent=2)}",
        f"## Agent Task\n{entry.agent_task}",
    ]
    if entry.score is not None or entry.tier is not None:
        parts.append(f"## Difficulty\nscore: {entry.score}, tier: {entry.tier}")
    return "\n\n".join(parts)


def build_phase_messages(
    entry: BenchmarkEntry,
    state: ReproState,
    phase: Phase,
) -> list[dict[str, Any]]:
    """Build the [system, user] message pair for a single phase."""
    system = AGENT_PREAMBLE + "\n\n" + phase.prompt_builder(entry, state) + (
        f"\n\nAllowed tools: {', '.join(phase.allowed_tools)}"
    )

    user = build_prompt(entry) + "\n\n" + state.state_summary()
    return [
        {"role": "system", "content": system},
        {"role": "user", "content": user},
    ]
